pkg/controller/kubedirectorcluster/util.go

```diff
@@ -92,52 +95,36 @@ func ClustersUsingApp(
 // ensureClusterAppReference notes that an app type is in use by this cluster.
 func ensureClusterAppReference(
 	cr *kdv1.KubeDirectorCluster,
-	handler *Handler,
+	reconciler *ReconcileKubeDirectorCluster,
 ) {
 	clusterKey := cr.Namespace + "/" + cr.Name
-	handler.lock.Lock()
-	defer handler.lock.Unlock()
-	handler.clusterState.clusterAppTypes[clusterKey] = cr.Spec.AppID
+	reconciler.lock.Lock()
+	defer reconciler.lock.Unlock()
+	reconciler.clusterState.clusterAppTypes[clusterKey] = cr.Spec.AppID
 }
 
 // removeClusterAppReference notes that an app type is no longer in use by
 // this cluster.
 func removeClusterAppReference(
-	cr *kdv1.KubeDirectorCluster,
-	handler *Handler,
+	namespacedName types.NamespacedName,
+	reconciler *ReconcileKubeDirectorCluster,
 ) {
-	clusterKey := cr.Namespace + "/" + cr.Name
-	handler.lock.Lock()
-	defer handler.lock.Unlock()
-	delete(handler.clusterState.clusterAppTypes, clusterKey)
-}
-
-// removeGlobalConfig removes the globalConfig from handler structure
-func removeGlobalConfig(handler *Handler) {
-	handler.lock.Lock()
-	defer handler.lock.Unlock()
-	handler.globalConfig = nil
-}
-
-// addGlobalConfig adds the globalConfig CR data to handler structure
-func addGlobalConfig(
-	handler *Handler,
-	cr *kdv1.KubeDirectorConfig,
-) {
-	handler.lock.Lock()
-	defer handler.lock.Unlock()
-	handler.globalConfig = cr
+	clusterKey := namespacedName.Namespace + "/" + namespacedName.Name
+	reconciler.lock.Lock()
+	defer reconciler.lock.Unlock()
+	delete(reconciler.clusterState.clusterAppTypes, clusterKey)
 }
 
 // getNativeSystemdSupport extracts the flag definition from globalConfig CR data
 // if present, otherwise returns false
 func getNativeSystemdSupport(
-	handler *Handler,
+	reconciler *ReconcileKubeDirectorCluster,
 ) bool {
-	handler.lock.RLock()
-	defer handler.lock.RUnlock()
-	if handler.globalConfig != nil && handler.globalConfig.Spec.NativeSystemdSupport != nil {
-		return *(handler.globalConfig.Spec.NativeSystemdSupport)
+	// Fetch global config CR (if present)
+	kdConfigCR, _ := observer.GetKDConfig(shared.KubeDirectorGlobalConfig, reconciler.Client)
+
+	if kdConfigCR != nil && kdConfigCR.Spec.NativeSystemdSupport != nil {
+		return *kdConfigCR.Spec.NativeSystemdSupport
```

Comment:
Eh. Hmm...

The only reason we watch the config object and cache it is so that we can then use the cached value, which this function now circumvents.

Looking in our codebase at the other function that uses the globalConfig -- the cluster validator -- I see that it currently does the same thing, always fetching the config from K8s and ignoring the cache. So that's not ideal either.

Anyway, we have the same sort of issue here that we do with clusterState, i.e. how to best share some data structures among the different kinds of handler processes. We could punt on that in the case of globalConfig and just always fetch it (and rip out its CR reconciliation handler entirely), but I don't think that will get rid of the general problem... in fact we will probably want to add even more cached stuff in the form of the app catalog, so that the cluster reconciler doesn't have to fetch the app definition CR each time it runs.

I think it's fine to have one or more global data structures that contain this shared data and the necessary locking facilities. We can look into whether we want to have those directly accessed wherever they are needed, or else accessed at just the very top-level entrypoints of the processes and then the necessary data structures passed down the call chain as arguments.

Reply:
We definitely need to discuss this offline. Once thing I will add though is that we're not actually fetching these objects from Kubernetes all the time. The new framework uses a split client that caches the objects and uses the cache a lot of the time.

Reply:
One more piece of info from the K8s documentation, "A split client reads (Get and List) from the Cache and writes (Create, Update, Delete) to the API server. Reading from the Cache significantly reduces request load on the API server; as long as the Cache is updated by the API server, read operations are eventually consistent."

Reply:
I moved the global data out of the KubeDirectorCluster controller to shared module so that it's cached and accessible from all the modules like it was before.